rllib/core/learner/learner.py

```diff
@@ -1010,192 +1011,139 @@ def update_from_batch(
+            # found in this batch. If not, throw an error.
+            unknown_module_ids = set(tensor_minibatch.policy_batches.keys()) - set(
+                self.module.keys()
+            )
+            if unknown_module_ids:
+                raise ValueError(
+                    f"Batch contains one or more ModuleIDs ({unknown_module_ids}) that "
+                    f"are not in this Learner!"
                 )
-                if len(unknown_module_ids) > 0:
-                    raise ValueError(
-                        "Batch contains one or more ModuleIDs that are not in this "
-                        f"Learner! Found IDs: {unknown_module_ids}"
-                    )
 
-                # Log metrics.
-                self._log_steps_trained_metrics(batch)
+            # Make the actual in-graph/traced `_update` call. This should return
+            # all tensor values (no numpy).
+            fwd_out, loss_per_module, tensor_metrics = self._update(
+                tensor_minibatch.policy_batches
+            )
 
-                # Make the actual in-graph/traced `_update` call. This should return
-                # all tensor values (no numpy).
-                fwd_out, loss_per_module, tensor_metrics = self._update(
-                    batch.policy_batches
-                )
-                # Convert logged tensor metrics (logged during tensor-mode of
-                # MetricsLogger) to actual (numpy) values.
-                self.metrics.tensors_to_numpy(tensor_metrics)
+            # Convert logged tensor metrics (logged during tensor-mode of MetricsLogger)
+            # to actual (numpy) values.
+            self.metrics.tensors_to_numpy(tensor_metrics)
 
-                self._set_slicing_by_batch_id(batch, value=False)
-                # If `num_iters` is reached break and return.
-                if num_iters and i == num_iters:
-                    break
+            # TODO (sven): Maybe move this into loop above to get metrics more accuratcely
+            #  cover the minibatch/epoch logic.
+            # Log all timesteps (env, agent, modules) based on given episodes/batch.
+            self._log_steps_trained_metrics(tensor_minibatch)
 
-        logger.debug(
-            f"===> [Learner {id(self)}] number of iterations run in this epoch: {i}"
-        )
+            self._set_slicing_by_batch_id(tensor_minibatch, value=False)
 
         # Log all individual RLModules' loss terms and its registered optimizers'
         # current learning rates.
+        # Note: We do this only once for the last of the minibatch updates, b/c the
+        # window is only 1 anyways.
         for mid, loss in convert_to_numpy(loss_per_module).items():
             self.metrics.log_value(
                 key=(mid, self.TOTAL_LOSS_KEY),
                 value=loss,
                 window=1,
             )
-        # Record the number of batches pulled from the dataset in this RLlib iteration.
```

Comment:
This is the location where these metrics should be collected. The `i` needs to be defined for this.

Reply:
See my other comment above. I think moving this to the above if-block (where we check for `training_data.data_iterator`) is actually accurate.